src/rust/engine/src/graph.rs

```diff
@@ -150,23 +150,25 @@ struct InnerGraph {
 
 impl InnerGraph {
   fn entry(&self, node: &EntryKey) -> Option<&Entry> {
-    self.entry_id(node).map(|&id| self.entry_for_id(id))
+    self.entry_id(node).and_then(|&id| self.entry_for_id(id))
   }
 
   fn entry_id(&self, node: &EntryKey) -> Option<&EntryId> {
     self.nodes.get(node)
   }
 
-  fn entry_for_id(&self, id: EntryId) -> &Entry {
-    self.pg.node_weight(id).unwrap_or_else(|| {
-      panic!("Invalid EntryId: {:?}", id)
-    })
+  fn entry_for_id(&self, id: EntryId) -> Option<&Entry> {
+    self.pg.node_weight(id)
   }
 
-  fn entry_for_id_mut(&mut self, id: EntryId) -> &mut Entry {
-    self.pg.node_weight_mut(id).unwrap_or_else(|| {
-      panic!("Invalid EntryId: {:?}", id)
-    })
+  fn entry_for_id_mut(&mut self, id: EntryId) -> Option<&mut Entry> {
+    self.pg.node_weight_mut(id)
+  }
+
+  fn unsafe_entry_for_id(&self, id: EntryId) -> &Entry {
+    self.pg.node_weight(id).expect(
+      "The unsafe_entry_for_id method should only be used in read-only methods!",
```

Comment:
maybe add a "while the graph lock is held" to the message?

Reply:
That's implied, since it's impossible to access these methods outside the lock.

Reply:
Aha, yes, this is on InnerGraph!

(Aside: The whole "key-safe HashMap" thing would actually be a very fancy standalone crate, if there's a way to make it that general...)